[Fuser] Do not attempt to use OpenMP if build without OpenMP support

malfet opened the pull request:
Clang from XCode does not support `-fopenmp` option, no need to try to compile with it.
Infer whether OpenMP is supported by checking _OPENMP define.
Also, use clang compiler if host app was compiled with clang rather than gcc.
Fix few range loop warnings and add static_asserts that range loop variables are raw pointers.

This changes makes fuser tests on OS X a bit faster.

Before:
```
% python3 test_jit.py -v  TestScript.test_batchnorm_fuser_cpu
Fail to import hypothesis in common_utils, tests are not derandomized
CUDA not available, skipping tests
test_batchnorm_fuser_cpu (__main__.TestScript) ... clang: error: unsupported option '-fopenmp'
clang: error: unsupported option '-fopenmp'
warning: pytorch jit fuser failed to compile with openmp, trying without it...
ok

----------------------------------------------------------------------
Ran 1 test in 0.468s

OK
```

After:
```
% python3 test_jit.py -v  TestScript.test_batchnorm_fuser_cpu
Fail to import hypothesis in common_utils, tests are not derandomized
CUDA not available, skipping tests
test_batchnorm_fuser_cpu (__main__.TestScript) ... ok

----------------------------------------------------------------------
Ran 1 test in 0.435s

OK
```

malfet committed:
[Fuser] Do not attempt to compile with OpenMP flags on Apple platform

Clang from XCode does not support `-fopenmp` option, no need to try to compile with it
Also, makes running fuser tests on OS X a bit faster.

Before:
```
% python3 test_jit.py -v  TestScript.test_batchnorm_fuser_cpu
Fail to import hypothesis in common_utils, tests are not derandomized
CUDA not available, skipping tests
test_batchnorm_fuser_cpu (__main__.TestScript) ... clang: error: unsupported option '-fopenmp'
clang: error: unsupported option '-fopenmp'
warning: pytorch jit fuser failed to compile with openmp, trying without it...
ok

----------------------------------------------------------------------
Ran 1 test in 0.468s

OK
```

After:
```
% python3 test_jit.py -v  TestScript.test_batchnorm_fuser_cpu
Fail to import hypothesis in common_utils, tests are not derandomized
CUDA not available, skipping tests
test_batchnorm_fuser_cpu (__main__.TestScript) ... ok

----------------------------------------------------------------------
Ran 1 test in 0.435s

OK
```

janeyx99 reviewed (approved):
interesting, looks fine

smessmer commented on torch/csrc/jit/codegen/fuser/cpu/fused_kernel.cpp:
```diff
@@ -196,11 +196,18 @@ struct CompilerConfig {
   const std::string openmp_flags = "-fopenmp";
 #endif
+#if defined(__APPLE__)
+  bool openmp = false;
```
can we add a comment for posterity on why we're doing this?

malfet replied:
Sure, OpenMP support is MacOS devtools is explicitly disabled:
```
   $ clang -c omp.c -fopenmp
   clang: error: unsupported option '-fopenmp'
```

malfet replied:
Updated the PR to rely less on the platform, but more on the compilation flags (i.e. openmp is set to false if _OPENMP flag is not defined)

malfet committed:
Update constraints to be a bit less platform specific

Use _OPENMP and __clang__ defines to detect whether host compiler is clang or gcc

malfet committed:
Fix two compilation warnings

Const-reference of a pointer is meaningless
Added a note and static_assert that type of range loop variable is a pointer
